### PR TITLE
multisig-sdk (incomplete. able to create multisig account.)

### PR DESCRIPTION
**Pull request by yenkhoon**
Request for review before continue.

### CODE REVIEW

File: src.ts/multisig.ts [hunk 1]
```diff
@@ -0,0 +1,306 @@
+'use strict';
+
+import * as errors from './errors';
+import { Provider, TransactionRequest, TransactionResponse, TransactionReceipt} from './providers/abstract-provider';
+import { Signer } from './abstract-signer';
+
+
```
Comment (jeansoonsik) [inline src.ts/multisig.ts]: remove extra line

File: src.ts/multisig.ts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+import { defineReadOnly, resolveProperties, checkProperties } from './utils/properties';
+import { populateTransaction } from './utils/transaction';
+
+import { checkFormat, checkString, checkNumber, checkAny } from './utils/misc';
+import { BigNumberish, Arrayish, deriveAddress } from './utils';
+
+
+
```
Comment (jeansoonsik) [inline src.ts/multisig.ts]: remove extra line

File: src.ts/multisig.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        }, true);
+
+        return resolveProperties({ signerAddress: signer.getAddress() }).then(({ signerAddress }) => {
+            if (!signerAddress) {
+                errors.throwError('create multisig wallet require signer address', errors.MISSING_ARGUMENT, { required: 'signer' });
+            }
+            properties.owner = signerAddress; // Set signer address as owner
+
+            let multisig: MultiSigWalletProperties = checkFormat({
+                owner: checkString,
```
Comment (jeansoonsik) [inline src.ts/multisig.ts]: It should be checkAddress?

File: src.ts/mxw.ts [hunk 4]
```diff
@@ -6,6 +6,7 @@ import * as errors from './errors';
 import { Signer } from './abstract-signer';
 import { Wallet } from './wallet';
 
+
```
Comment (jeansoonsik) [inline src.ts/mxw.ts]: remove extra line

File: src.ts/multisig.ts [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+                threshold: checkNumber,
+                signers: checkAny,
```
Comment (jeansoonsik) [inline src.ts/multisig.ts]: The `signers` should be a list of address. Should we use the `arrayOf(checkAddress)`?

File: src.ts/multisig.ts [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+            let transaction = signer.provider.getTransactionRequest("multisig", "auth-createMultiSigAccount", {
+                from: multisig.owner,
+                threshold: multisig.threshold,
+                signers: multisig.signers
+            });
+            transaction.fee = signer.provider.getTransactionFee(undefined, undefined, { tx: transaction });
+            
+            return signer.sendTransaction(transaction, overrides).then((response) => {
+                
+                let groupAddress = deriveAddress(signerAddress, signer.getNonce());
```
Comment (jeansoonsik) [inline src.ts/multisig.ts]: The calculation for the groupAddress will be incorrect when the `bulkSend` flag is enable during the sendTransaction, the nonce will be increase by one internally. It should be calculate before sendTransaction.

File: src.ts/multisig.ts [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+        return resolveProperties({ address: signer.getAddress() }).then(({ address }) => {
+            if (!address) {
+                errors.throwError('update multisig wallet require signer address', errors.MISSING_ARGUMENT, { required: 'signer' });
+            }
+            properties.from = address; // Set signer address as owner
+
+            let multisig: UpdateMultiSigWalletProperties = checkFormat({
+                from: checkString,
+                groupAddress: checkString,
```
Comment (jeansoonsik) [inline src.ts/multisig.ts]: It should be `checkAddress`

File: src.ts/multisig.ts [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+                threshold: checkNumber,
+                signers: checkString,
```
Comment (jeansoonsik) [inline src.ts/multisig.ts]: The `signers` should be a list of address. Should we use the `arrayOf(checkAddress)`?

**Comment by jeansoonsik** [comment]
Please pull request from a feature branch

File: src.ts/multisig.ts (hunk 8 above)
Comment (yenkhoon) [reply src.ts/multisig.ts]: fxied

File: src.ts/multisig.ts (hunk 7 above)
Comment (yenkhoon) [reply src.ts/multisig.ts]: fixed

**Pull request closed without merging**